### PR TITLE
Issue 4 item read

### PR DESCRIPTION
**Pull request by evtimo**
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Evgeny <zheny.tim@gmail.com>

### CODE REVIEW

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 1]
```diff
@@ -233,7 +243,40 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
...
+		}
+
+		try (final ClientFactory clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+		     EventStreamReader<String> reader = clientFactory.createReader("reader",
+				     readerGroup,
+				     new JavaSerializer<String>(),
+				     ReaderConfig.builder().build())) {
+			System.out.format("Reading all the events from %s/%s%n", scopeName, streamName);
+			EventRead<String> event = null;
+			do {
```
Comment (veronikaKochugova) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: uniformity is welcomed :+1: It is better to replace on "while(..) { //do}"

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 2]
```diff
@@ -254,15 +297,15 @@ private boolean invokePathRead(final PathOperation<? extends PathItem> pathOp) {
 		}
 
 		try (final ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
-			 EventStreamReader<String> reader = clientFactory.createReader("reader",
-					 readerGroup,
-					 new JavaSerializer<String>(),
-					 ReaderConfig.builder().build())) {
+		     EventStreamReader<String> reader = clientFactory.createReader("reader",
```
Comment (veronikaKochugova) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: final

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 3]
```diff
@@ -254,15 +297,15 @@ private boolean invokePathRead(final PathOperation<? extends PathItem> pathOp) {
...
+				     readerGroup,
+				     new JavaSerializer<String>(),
+				     ReaderConfig.builder().build())) {
 			System.out.format("Reading all the events from %s/%s%n", scope, streamName);
 			EventRead<String> event = null;
 			do {
```
Comment (veronikaKochugova) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: too hard. Why not like this:
`
EventRead<String> event;`

`while(null!= (event = reader.readNext(...))){//do}`

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 4]
```diff
@@ -254,15 +297,15 @@ private boolean invokePathRead(final PathOperation<? extends PathItem> pathOp) {
...
 				try {
-					event = reader.readNextEvent(readerTimeoutMs);
+					event = reader.readNextEvent(Constants.READER_TIMEOUT_MS);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: I thought that the configuration option has been merged already. Why does the constant is used here instead?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 5]
```diff
@@ -170,7 +173,14 @@ private void invokeFileNio(DataOperation<? extends DataItem> fileOperation) {
 					//TODO: EventStreamWriter<ByteBuffer>  ->  writeEvent
 					break;
 				case READ:
-					//TODO: EventStreamReader<ByteBuffer>  ->  readNextEvent
+					if ((null != scopeMap.get(scopeName) && null != scopeMap.get(scopeName).get(streamName))) {
+						if (invokeDataRead(fileOperation, scopeName, streamName)) {
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:180]: looks like something should be done if condition is `true`

**Review by akurilov — CHANGES_REQUESTED** [review]

**Review by veronikaKochugova — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 1 above)
Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: may be replaced like this:
`for(event = reader.readNextEvent(...); null != event.getEvent();) {`

also we should determine the size of the event been read and update the load operations status

**Review by akurilov — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 5 above)
Comment (evtimo) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:180]: nothing should be done, DataRead is caused in condition.
we can make some logs if returns TRUE

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 3 above)
Comment (evtimo) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: so, will use 'for' loop, as Andrey mentioned

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 5 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:180]: we agreed that we don't this check. Do we need it now?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 6]
```diff
@@ -173,7 +176,14 @@ private void invokeFileNio(DataOperation<? extends DataItem> fileOperation) {
...
+						}
+					} else {
+						Loggers.ERR.debug(
+								"Failed to read the stream {} in the scope {}", streamName, scopeName);
```
Comment (dadlex) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:184]: as we discussed with Andrey, we don't need to check scope and stream before reading. We just need to correctly react if pravega falls due to absent scope or stream.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 5 above)
Comment (evtimo) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:180]: If first condition is FALSE, the second will be never executed.
So, there will be no exception. (as I understand)
"short-circuiting" behavior in Java

Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:180]: We don't need the check with maps at all.

**Pull request closed without merging**